tests/cache_test.py

```diff
@@ -172,3 +171,40 @@ def lambda_handler(event, context):
             return 'Hello from Lambda with secret %s' % secret_value
         
         lambda_handler(None, None)
+
+    # This isn't any different than the default behaviour, but allows
+    # us to confirm that the code calls the provided factory method.
+    def _get_client(self, session, name):
```

Comment:
Could you please turn these two utility methods into actual mocks?

Something like the following:

```python
from mock import MagicMock
session_factory = MagicMock(return_value=boto3.session.Session())
client_factory = MagicMock(return_value=session_factory().client('ssm'))
```

then you can test it:

```python
client_factory.assert_called()  # or even .assert_called_once()
```

This way you will avoid tricks such as `self.got_session`.

Reply:
Definitely! Thanks! (it's probably pretty obvious that Python is not my primary language.)

Reply:
I ran into some issues trying to use `MagicMock` -- it was complaining about `__name__`. I realized after poking at it for a bit that the logic I was trying to test with the mocks is tested implicitly in pretty much everything else, so I've replaced the test with a different one that tests the code path where the factory variables are set to `None`.